### ISSUE TITLE
TokenMgrError is a subclass of Error instead of Exception

### ISSUE THREAD

**Issue by SerialVelocity** [JSQLParser/JSqlParser#1398]
**Describe the bug**
An error is thrown instead of an exception when a token parsing error occurs.

An error should only really be thrown for very serious errors like out of memory, thread death, etc. Even an NPE is only classes as an exception. If you should be catching it, it shouldn't be an Error, it should be an exception.

From the Java docs:
> An Error is a subclass of Throwable that indicates serious problems that a reasonable application should not try to catch. Most such errors are abnormal conditions. The ThreadDeath error, though a "normal" condition, is also a subclass of Error because most applications should not try to catch it.

**To Reproduce**
1. Example SQL: Unknown as our server shutsdown instead of catching it (as should happen with Errors).
2. Parsing this SQL using JSqlParser with this statements
3. Exception:
```
net.sf.jsqlparser.parser.TokenMgrError: Lexical error at line 4, column 39.  Encountered: "\u00a8" (168), after : ""
	at net.sf.jsqlparser.parser.CCJSqlParserTokenManager.getNextToken(CCJSqlParserTokenManager.java:5943)
```

**Expected behavior**
An exception to be thrown instead of an error.

**System**
 - Database you are using Postgres
- Java Version 15
- JSqlParser version 4.3-SNAPSHOT

**Comment by manticore-projects** [comment]
Greetings.

Pardon me to be blunt here: In my understanding, JSQLParser parses your statement or it does not -- and the error message will be helpful only when you are extremely lucky. I can easily create statements, where the error is thrown at irrelevant positions and with misguiding hints (especially when sub-queries are involved).

So as far as I can see it, your concern is a bit academical and if you really want to change the Error into an Exception then you would need to send a PR or sponsor a developer.

**Comment by SerialVelocity** [comment]
So, my problem isn't the message that JSQLParser outputs, my problem is the type of the thrown object.

You can see someone has encountered the problem here in the past:
https://github.com/JSQLParser/JSqlParser/issues/31#issuecomment-358998915

Generally for good code you would only catch exceptions and let errors propagate:
```
try {
  // parse query here
} catch (Exception e) {
  // Handle exception
}
```
You could even make it `catch (RuntimeException | JSQLParserException e)`

Currently, that doesn't work as you are throwing an `Error`, not an `Exception`. Most people don't want to catch `Error` or `Throwable` because you might end up catching `OutOfMemoryError`, etc which you want the program to fail when that happens.

Currently, you need (assuming you aren't throwing any more error types):
```
try {
  // parse query here
} catch (Exception | Error e) {
  if (e instanceof Error && !(e instanceof TokenMgrError)) {
    // Rethrow all important errors
    throw e;
  }
  // Handle exception
}
```

Does that make sense?

It just seems odd that a bad input character could take down a server.

**Comment by manticore-projects** [comment]
Sorry, I am not able to follow. For the JSQLFormatter Online Service, we basically use oce like shown below:
```java
try {
Statement statement =
              CCJSqlParserUtil.parse(
                  statementSql,
                  parser -> parser.withSquareBracketQuotation(useSquareBracketQuotation));
} catch (Exception ex) {

}
```
This server runs since forever, while it is just a small shared memory starved instance.
What would be wrong with that from your point of view? And why do you think the parser would take a server down?

**Comment by manticore-projects** [comment]
> ```
> SELECT p as ¨P FROM dual
> ```

Thank you for the example, I start to see your point.
[When entering the statement above into JSQLFormatterUI](http://jsqlformatter.manticore-projects.com/jsqlformatter/demo.html?args=-c%20MoUQMiDCAqAEAOsCGBnWAKgCrAYgJQHkBZWAEwFckAbIA+), it will not show an Parser `Exception`, but throws an `Error` when trying to format it.

I will likely have a look into this over the weekend and would return with findings.

**Comment by SerialVelocity** [comment]
Hi Andreas! Did you manage to take a look in the end?

I just tried running that query on 4.2 and realised that it is handled correctly in that version:
```
net.sf.jsqlparser.JSQLParserException: Lexical error at line 1, column 13.  Encountered: "\u00a8" (168), after : ""
	at net.sf.jsqlparser.parser.CCJSqlParserUtil.parseStatement(CCJSqlParserUtil.java:190)
	at net.sf.jsqlparser.parser.CCJSqlParserUtil.parse(CCJSqlParserUtil.java:63)
	at net.sf.jsqlparser.parser.CCJSqlParserUtil.parse(CCJSqlParserUtil.java:38)
```

So it looks like it is a regression in 4.3-SNAPSHOT. I had a look at the changes and I couldn't find the change that caused this class to show up. Do you happen to know which PR/lines introduced this new TokenMgrError class?

**Comment by manticore-projects** [comment]
Hi Ben,

sorry, I have had a hell of a week and not much time on my hands.
I do not think, anything has changed on the Token Manager itself, but new Tokens and Productions have been added. It is very possible/likely that your statement was completely rejected before while now it is partially parsed and then fails hard (or the other way around).

My intention is to implement a test tool, which iterates through each GIT commit, compiles and tests and reports output. I will need that for the performance discussion also. Although I can't promise, just sketch.

**Comment by SerialVelocity** [comment]
No problem, I expected something like that :)

It's a bit weird that the exception message is the same apart from the type of exception! You mentioning partially parsed reminds me that I saw some `LOOKAHEAD` stuff added so maybe that is caused the new error class to be generated.

That test tool sounds valuable for performance testing! For finding out when issues like this were introduced, `git blame` tends to be a bit faster since it does a binary search for the commit.

**Comment by wumpz** [comment]
Could this again a difference to the gradle build? Maven build does not generate a TokenMgrError class but a TokenMgrException class.

**Comment by wumpz** [comment]
This `TokenMgrError` is generated when the legacy error handling of JavaCC is activated:
 https://github.com/javacc/javacc/blob/e6620f210ce182af53d09ca1a556aba82f4cc244/src/main/java/org/javacc/parser/JavaFiles.java#L488

Since 08.2019 we have this option activated

`JAVA_TEMPLATE_TYPE = "modern";`

This one seems not to be honored using the gradle build.

**Comment by wumpz** [comment]
As for the tools look into: https://github.com/JSQLParser/javacc-perf-diag. This one make a statistics about the parsing using aspects that are connected to the parser.

**Comment by manticore-projects** [comment]
On Fri, 2021-11-19 at 15:43 -0800, Tobias wrote:
> This TokenMgrError is generated when the legacy error handling of
> JavaCC is activated:
> https://github.com/javacc/javacc/blob/e6620f210ce182af53d09ca1a556aba82f4cc244/src/main/java/org/javacc/parser/JavaFiles.java#L488
> Since 08.2019 we have this option activated
> JAVA_TEMPLATE_TYPE = "modern";
> This one seems not to be honored using the gradle build.

You are right.
Adding the dependency: javacc 'net.java.dev.javacc:javacc:7.0.10'
should solve this problem though. I would submit a PR soonest.

**Comment by manticore-projects** [comment]
Solved by PR #1428 which requests for latest JavaCC 7.0.10 explicitely.
